### PR TITLE
Layer order bugfix [WIP]

### LINKED ISSUE #7621: Ordering layers is not working in this case
## STR

Making the example with this map: https://nycdoittpublicdata.cartodb.com/u/nycpublicdata/viz/be21c8c2-def6-11e5-b40b-0e787de82d45/public_map.

1. Export and import the map in your account.
2. Go to the map (current editor) and change the basemap (just for commodity) to one without labels.
3. Go to the new builder.
3. Open the dev chrome console, we will need it.
4. Add a widget for `nyc_subway_line`, e.g `cartodb_id` category.
5. Go back to layers list and move layer layer `C`, `nyc_subway_line` to the last position.
6. Check that tiles have not been reloaded.
7. Get back `C` layer to first position.
8. Tiles have been reloaded but widget, suddenly, is not working.

cc @alonsogarciapablo

### PR DESCRIPTION
**Pull request by xavijam**
Two things:

1. We were sending incorrect orders due to `resetOrders` happens after the `add` action who starts all the machinery.
2. When that new `add` event provokes a new createLayer (:ok_hand:), but in that method we remove the source, that was the reason why we were losing the analysis definition in the map instantiation.

What do you think @alonsogarciapablo?
Fixes #7621

### CODE REVIEW

File: lib/assets/javascripts/cartodb3/deep-insights-integrations.js [hunk 1]
```diff
@@ -183,7 +184,9 @@ var LAYER_TYPE_TO_LAYER_CREATE_METHOD;
 F.prototype._createLayer = function (m, opts) {
   opts = opts || {};
   var attrs = JSON.parse(JSON.stringify(m.attributes)); // deep clone;
-  delete attrs.source; // a new layer should not have an analyse to begin with, even if we have it in the context of editor
+  if (!opts || !opts.keepSource) {
+    delete attrs.source; // a new layer should not have an analyse to begin with, even if we have it in the context of editor
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/deep-insights-integrations.js]: I'm wondering if could check if there's an existing analysisDefinition for attrs.source and, if that's the case, we would not need to `delete attrs.source` cause we can assume the analysis definition node exists in the context of CartoDB.js.

@xavijam @viddo what do you think?

Comment (viddo) [inline lib/assets/javascripts/cartodb3/deep-insights-integrations.js]: Sounds reasonable 👍

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/deep-insights-integrations.js]: Well, there will be always an analysis for each layer (the source one).

Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/deep-insights-integrations.js]: Yes, there will always be an analysis node... What I'm suggesting is checking it there's an existing **analysisDefinition** whose `node_id` points to `attrs.source` in the analysisDefinitionsCollection...

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/deep-insights-integrations.js]: Ahhhhh, fair enough!

Comment (viddo) [inline lib/assets/javascripts/cartodb3/deep-insights-integrations.js]: 👍  `this._analysisDefinitionsCollection.findByNodeId(m.get('source'))` should do :)

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js [hunk 2]
```diff
@@ -47,9 +47,10 @@ module.exports = Backbone.Collection.extend({
 
     var movingLayer = this.at(from);
     this.remove(movingLayer, { silent: true });
-    this.add(movingLayer, { at: to, parse: false });
+    this.add(movingLayer, { at: to, parse: false, silent: true });
     this._resetOrders();
     this.save();
+    this.trigger('add', movingLayer, this, { at: to });
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: We should probably trigger this event after `this.save()` have succeeded, right?

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js [hunk 3]
```diff
@@ -47,9 +47,10 @@ module.exports = Backbone.Collection.extend({
 
     var movingLayer = this.at(from);
     this.remove(movingLayer, { silent: true });
-    this.add(movingLayer, { at: to, parse: false });
+    this.add(movingLayer, { at: to, parse: false, silent: true });
```
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: I know this wasn't changed now, but I'm a bit skeptical about the parse: false. Why not maintain the model and just update the order attr? Not doing the parse could have unwanted side-effects (e.g. internal models not being setup/initialized as expected).

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: If we agree with that, I'd open a new ticket about that issue. I don't want to mix things and make the result bigger.
But in any case, if the solution was reviewed, we should keep with the idea, shouldn't we?

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 2 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: :+1:

File: lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js (hunk 3 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: > …, if the solution was reviewed, we should keep with the idea, shouldn't we?

Why? That something was reviewed previously doesn't mean that it's correct. 
But +1 to open new issue to look into this one.

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js]: Totally agree, let's open a new ticket.

**Comment by alonsogarciapablo** [comment]
:+1:

File: lib/assets/test/spec/cartodb3/data/layer-definitions-collection.spec.js [hunk 4]
```diff
@@ -499,16 +499,22 @@ describe('data/layer-definitions-collection', function () {
...
+      };
       this.collection.on('add', onAddCallback);
       this.collection.on('remove', onRemoveCallback);
+      this.collection.on('layerMoved', onLayerMovedCallback);
 
       this.collection.moveLayer({ from: 3, to: 1 });
 
-      expect(onAddCallback).toHaveBeenCalled();
+      expect(onAddCallback).not.toHaveBeenCalled();
+      expect(onLayerMovedCallback).toHaveBeenCalled();
```
Comment (alonsogarciapablo) [inline lib/assets/test/spec/cartodb3/data/layer-definitions-collection.spec.js]: `expect(onLayerMovedCallback).toHaveBeenCalledWith(this.collection.at(x), 1);`?

File: lib/assets/javascripts/cartodb3/deep-insights-integrations.js [hunk 5]
```diff
@@ -189,8 +189,8 @@ F.prototype._onLayerDefinitionRemoved = function (m) {
 };
 
 F.prototype._onLayerDefinitionMoved = function (m, at, c) {
-  if (!m) throw new Error('moved layer model not provided');
-  if (!at) throw new Error('at position provided');
+  if (!m) throw new Error('moved layer model required');
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/deep-insights-integrations.js]: I'd personally not include this checks here. Instead, I'd ensure that whoever triggers a "layerMoved" event passes the right arguments (as suggested [here](https://github.com/CartoDB/cartodb/pull/7623/commits/373a7082bf975330a236d0e42816dd85dbe9f8b0#r65712758))

File: lib/assets/test/spec/cartodb3/deep-insights-integrations.spec.js [hunk 6]
```diff
@@ -286,4 +286,37 @@ describe('deep-insights-integrations', function () {
       });
     });
   });
+
+  describe('when a layer is moved', function () {
+    beforeEach(function () {
+      // this.analysisDefinitionsCollection.add({ node_id: 'a0' });
```
Comment (alonsogarciapablo) [inline lib/assets/test/spec/cartodb3/deep-insights-integrations.spec.js]: We have some commented out here. Just sayin'

**Comment by alonsogarciapablo** [comment]
Left a couple of super-minor comments (feel free to ignore them)

 :+1: :+1: :+1:

**Pull request merged**